### PR TITLE
NVSHAS-8857-support image pagination

### PR DESCRIPTION
**Pull request by jeffhuang4704**

### CODE REVIEW

File: controller/access/access.go [hunk 1]
```diff
@@ -759,7 +759,7 @@ func CompileUriPermitsMapping() {
 				"v1/scan/platform/platform",
 				"v1/scan/asset",
 				"v1/vulasset",
-				// "scanasset5", // TODO: local dev only
+				"v1/scan/asset/images",
```
Comment (williamlin-suse) [inline controller/access/access.go:762]: Please update access/access_test.go & Permission_API.txt for this new API (see "/v1/vulasset" for reference)

File: controller/access/access.go [hunk 2]
```diff
@@ -951,8 +951,7 @@ func CompileUriPermitsMapping() {
 				"v1/scan/platform/platform",
 				"v1/vulasset",
 				"v1/assetvul",
-				// "scanasset5",     // TODO: local dev only
-				// "scanassetview1", // TODO: local dev only
+				"v1/scan/asset/images",
```
Comment (williamlin-suse) [inline controller/access/access.go:954]: Please update access/access_test.go & Permission_API.txt for this new API (see "v1/assetvul" for reference)

File: controller/rest/vulasset.go [hunk 3]
```diff
@@ -466,14 +545,155 @@ func combineQueryFilter(r *http.Request) (*db.VulQueryFilter, error) {
...
+
+	restRespSuccess(w, r, queryStat, acc, login, nil, "Create asset session")
+
+	// populate the result to file-based database
+	go func() {
+		err = db.DupAssetSessionTableToFile(queryFilter.QueryToken)
+		if err != nil {
+			log.WithFields(log.Fields{"err": err}).Error("PopulateSessionToFile error")
+		}
+	}()
```
Comment (holyspectral) [inline controller/rest/vulasset.go:643]: Question: Is there a reason why this has to be deferred to a go routine?

**Review by holyspectral — COMMENTED** [review]
Thanks for working on this.  It generally looks good to me.  I like the refactoring which makes existing function can be used by the new logics.

A quick summary of my comments:

1. Please increase `QueryToken` to be longer, at least 16 bytes.  It's a cheap fix but can avoid issues in the future, since collision can lead to information leakage.  
2. Define an enum to replace `0` and `1` in db.QueryStat.Type. 
3. Others are either questions, minor or optional.  

It's not must-to-have to me, but like William mentioned, some unit-tests would be fantastic.

**Review by jayhuang-suse — COMMENTED** [review]
Good. Please resolve the other reviewer's comments.

File: controller/access/access.go (hunk 1 above)
Comment (jeffhuang4704) [reply controller/access/access.go:762]: added in access/access_test.go.
added in Permission_API.txt

File: controller/access/access.go (hunk 2 above)
Comment (jeffhuang4704) [reply controller/access/access.go:954]: added in access/access_test.go.
added in Permission_API.txt

File: controller/cache/scan.go [hunk 4]
```diff
@@ -531,7 +531,10 @@ func scanDone(id string, objType share.ScanObjectType, report *share.CLUSScanRep
 			}
 		}
 
-		db.PopulateAssetVul(dbAssetVul)
+		err := db.PopulateAssetVul(dbAssetVul)
+		if err != nil {
+			log.WithFields(log.Fields{"err": err}).Error("Failed to poulate asset to db")
```
[the remark this answers is not in the document]
Comment (jeffhuang4704) [reply controller/cache/scan.go:536]: fixed.

File: controller/rest/scanner.go [hunk 5]
```diff
@@ -865,3 +865,17 @@ func handlerScanCacheData(w http.ResponseWriter, r *http.Request, ps httprouter.
 		restRespSuccess(w, r, resp, acc, login, nil, "Get scan cache data")
 	}
 }
+
+func handlerAssetViewCreate(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	if r.Method == http.MethodPost {
```
[the remark this answers is not in the document]
Comment (jeffhuang4704) [reply controller/rest/scanner.go:870]: fixed.

File: controller/rest/scanner.go [hunk 6]
```diff
@@ -865,3 +865,17 @@ func handlerScanCacheData(w http.ResponseWriter, r *http.Request, ps httprouter.
...
+		createAssetSession(w, r)
+		return
+	}
+}
+
+func handlerAssetViewGet(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	if r.Method == http.MethodGet {
```
[the remark this answers is not in the document]
Comment (jeffhuang4704) [reply controller/rest/scanner.go:877]: fixed.

File: controller/rest/vulasset.go (hunk 3 above)
Comment (jeffhuang4704) [reply controller/rest/vulasset.go:643]: for better performance as this is a time consuming task and it's inside the http request flow.

File: controller/rest/vulasset.go [hunk 7]
```diff
@@ -48,6 +49,23 @@ func createVulAssetSessionV2(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		log.WithFields(log.Fields{"err": err}).Error("createQueryStat error")
+		restRespErrorMessage(w, http.StatusInternalServerError, api.RESTErrInvalidRequest, err.Error())
+		return
+	}
+
+	// put to Consul to signal other controllers to build the same query session
+	err = createRequestInConsul(acc, login, 0, queryFilter.QueryToken, queryFilter.Filters, nil)
+	if err != nil {
+		log.WithFields(log.Fields{"err": err}).Error("Write create session request to consul error")
```
[the remark this answers is not in the document]
Comment (jeffhuang4704) [reply controller/rest/vulasset.go:66]: fixed.

File: controller/rest/vulasset.go [hunk 8]
```diff
@@ -356,6 +377,64 @@ func _createQuerySession(qsr *api.QuerySessionRequest) error {
...
+	}
+
+	allowed := getAllAllowedResourceId(acc)
+
+	_, _, err := db.CreateImageAssetSession(allowed, queryFilter)
+	if err != nil {
+		return err
+	}
+
+	queryFilterb, _ := json.Marshal(queryFilter)
```
[the remark this answers is not in the document]
Comment (jeffhuang4704) [reply controller/rest/vulasset.go:403]: fixed.

File: controller/scan/registry.go [hunk 9]
```diff
@@ -1911,16 +1916,43 @@ func IsRegistryImageScanned(id string) bool {
 }
 
 func getImageDbAssetVul(c *imageInfoCache, sum *share.CLUSRegistryImageSummary, criticals, highs, meds, lows []string) *db.DbAssetVul {
+	b, _ := json.Marshal(sum.Images)
```
[the remark this answers is not in the document]
Comment (jeffhuang4704) [reply controller/scan/registry.go:1919]: fixed.

File: controller/rest/vulasset.go [hunk 10]
```diff
@@ -48,6 +49,23 @@ func createVulAssetSessionV2(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// save the data to querystat table (do this before put request to Consul)
+	queryFilter.QueryToken = utils.GetRandomID(6, "") // do not change the length
+
+	queryFilterb, _ := json.Marshal(queryFilter)
```
[the remark this answers is not in the document]
Comment (jeffhuang4704) [reply controller/rest/vulasset.go:55]: fixed

File: controller/api/apis.go [hunk 11]
```diff
@@ -4042,14 +4051,42 @@ type RESTImageAssetView struct {
...
+	Name      string `json:"repository"`
+	Critical  int    `json:"critical"`
+	High      int    `json:"high"`
+	Medium    int    `json:"medium"`
+	CreatedAt string `json:"created_at"`
+	ScannedAt string `json:"scanned_at"`
+	Digest    string `json:"digest"`
+	BaseOS    string `json:"base_os"`
+	RegName   string `json:"reg_name"`
+	Registry  string `json:"repo_url"`
```
[the remark this answers is not in the document]
Comment (jeffhuang4704) [reply controller/api/apis.go:4065]: yes, checked with UI team.

File: controller/rest/vulasset.go [hunk 12]
```diff
@@ -48,6 +49,23 @@ func createVulAssetSessionV2(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// save the data to querystat table (do this before put request to Consul)
+	queryFilter.QueryToken = utils.GetRandomID(6, "") // do not change the length
```
[the remark this answers is not in the document]
Comment (jeffhuang4704) [reply controller/rest/vulasset.go:53]: create a case to follow up. 
https://jira.suse.com/browse/NVSHAS-9250

File: controller/rest/vulasset.go [hunk 13]
```diff
@@ -466,14 +545,155 @@ func combineQueryFilter(r *http.Request) (*db.VulQueryFilter, error) {
...
+		return
+	}
+
+	if queryFilter.Filters.Type != "image" {
+		restRespErrorMessage(w, http.StatusBadRequest, api.RESTErrInvalidRequest, "invalid asset type")
+		return
+	}
+
+	// create a query session
+	queryFilter.QueryToken = utils.GetRandomID(6, "") // do not change the length
```
[the remark this answers is not in the document]
Comment (jeffhuang4704) [reply controller/rest/vulasset.go:587]: create a case to follow up.
https://jira.suse.com/browse/NVSHAS-9250

File: controller/rest/vulasset.go [hunk 14]
```diff
@@ -331,7 +352,7 @@ func _createQuerySession(qsr *api.QuerySessionRequest) error {
 		LoginID:      qsr.UserAccess.LoginID,
 		LoginName:    qsr.UserAccess.LoginName,
 		Data1:        string(queryFilterb),
-		Data2:        string(perfStatsb),
+		Type:         0,
```
[the remark this answers is not in the document]
Comment (jeffhuang4704) [reply controller/rest/vulasset.go:355]: fixed.

File: db/vulassets.go [hunk 15]
```diff
@@ -830,6 +888,20 @@ func fillCVERecordV2(record *DbVulAsset) error {
...
+			"type": assetType,
+		}
+
+		part2_allowed := goqu.Ex{}
+		if len(allowedID) > 0 {
+			part2_allowed = goqu.Ex{
+				"assetid": allowedID,
+			}
+		}
+		return goqu.And(part1_type, part2_allowed)
```
[the remark this answers is not in the document]
Comment (jeffhuang4704) [reply db/vulassets.go:902]: it will become select all

**Comment by jeffhuang4704** [comment]
close this PR as I have a replacement - https://github.com/neuvector/neuvector/pull/1473

**Pull request closed without merging**